src/internal/resolver.ts

```diff
@@ -0,0 +1,461 @@
+    sourceName: string,
+    absolutePath: string,
+    libraryName?: string,
+    libraryVersion?: string,
+  ): Promise<ResolvedFile> {
+    const rawContent = await this._readFile(absolutePath);
+    const stats = await fsExtra.stat(absolutePath);
+    const lastModificationDate = new Date(stats.ctime);
+
+    const contentHash = createNonCryptographicHashBasedIdentifier(Buffer.from(rawContent)).toString("hex");
+
+    const parsedContent = this._parser.parse(rawContent, absolutePath, contentHash);
+
+    const content = {
+      rawContent,
+      ...parsedContent,
+    };
+
+    return new ResolvedFile(
+      sourceName,
+      absolutePath,
+      content,
+      contentHash,
+      lastModificationDate,
+      libraryName,
+      libraryVersion,
+    );
+  }
+
+  private _isRelativeImport(from: ResolvedFile, imported: string): boolean {
+    return (
+      imported.startsWith("./") ||
+      imported.startsWith("../") ||
+      fsExtra.existsSync(this._getImportAbsolutePath(from.absolutePath, imported))
+    );
+  }
+
+  private _getImportAbsolutePath(fromAbsolutePath: string, imported: string): string {
+    return fromAbsolutePath.replace(CIRCOM_FILE_REG_EXP, imported);
+  }
+
+  private _resolveNodeModulesFileFromProjectRoot(fileName: string) {
+    return resolve.sync(fileName, {
+      basedir: this._projectRoot,
+      preserveSymlinks: true,
+    });
+  }
+
+  private _getLibraryName(sourceName: string): string {
+    let endIndex: number;
+    if (this._isScopedPackage(sourceName)) {
+      endIndex = sourceName.indexOf("/", sourceName.indexOf("/") + 1);
+    } else if (sourceName.indexOf("/") === -1) {
+      endIndex = sourceName.length;
+    } else {
+      endIndex = sourceName.indexOf("/");
+    }
+
+    return sourceName.slice(0, endIndex);
+  }
+
+  private _getUriScheme(s: string): string | undefined {
+    const re = /([a-zA-Z]+):\/\//;
+    const match = re.exec(s);
+    if (match === null) {
+      return undefined;
+    }
+
+    return match[1];
+  }
+
+  private _isInsideSameDir(sourceNameInDir: string, sourceNameToTest: string) {
+    const firstSlash = sourceNameInDir.indexOf("/");
+    const dir = firstSlash !== -1 ? sourceNameInDir.substring(0, firstSlash) : sourceNameInDir;
+
+    return sourceNameToTest.startsWith(dir);
+  }
+
+  private _isScopedPackage(packageOrPackageFile: string): boolean {
+    return packageOrPackageFile.startsWith("@");
+  }
+
+  private _isRelativeImportToLibrary(from: ResolvedFile, imported: string): boolean {
+    return (
+      this._isRelativeImport(from, imported) && from.library === undefined && imported.includes(`${NODE_MODULES}/`)
+    );
+  }
+
+  private _relativeImportToLibraryToSourceName(from: ResolvedFile, imported: string): string {
+    const sourceName = normalizeSourceName(path.join(path.dirname(from.sourceName), imported));
+
+    const nmIndex = sourceName.indexOf(`${NODE_MODULES}/`);
+    return sourceName.substr(nmIndex + NODE_MODULES.length + 1);
```

Comment:
`substr` is deprecated

```suggestion
    return sourceName.substring(nmIndex + NODE_MODULES.length + 1);
```